silicon/backends/rabbitmq.hh

```diff
@@ -0,0 +1,260 @@
+# include <amqp.h>
+# include <amqp_framing.h>
+
+# include <iod/json.hh>
+
+# include <silicon/symbols.hh>
+# include <silicon/service.hh>
+
+iod_define_symbol(hostname)
+iod_define_symbol(exchange)
+iod_define_symbol(bindingkey)
+iod_define_symbol(username)
+iod_define_symbol(password)
+
+namespace sl
+{
+namespace rmq
+{
+	auto get_string = [] (auto const & b) { return std::string(static_cast<char const *>(b.bytes), b.len); };
+
+	void
+	die_on_error(int x, std::string const & context)
+	{
+		if (x < 0)
+			throw std::runtime_error(context + ": " + amqp_error_string2(x));
+	}
+
+	void
+	die_on_amqp_error(amqp_rpc_reply_t const & x, std::string const & context)
+	{
+		switch (x.reply_type)
+		{
+			case AMQP_RESPONSE_NORMAL:
+				return;
+
+			case AMQP_RESPONSE_NONE:
+				throw std::runtime_error(context + ": missing RPC reply type!");
+
+			case AMQP_RESPONSE_LIBRARY_EXCEPTION:
+				throw std::runtime_error(context + ": " + amqp_error_string2(x.library_error));
+
+			case AMQP_RESPONSE_SERVER_EXCEPTION:
+				switch (x.reply.id)
+				{
+					case AMQP_CONNECTION_CLOSE_METHOD:
+					{
+						auto m = static_cast<amqp_connection_close_t const *>(x.reply.decoded);
+
+						throw std::runtime_error(context + ": server connection error " + std::to_string(m->reply_code)
+								+ ", message: " + get_string(m->reply_text));
+					}
+
+					case AMQP_CHANNEL_CLOSE_METHOD:
+					{
+						auto m = static_cast<amqp_channel_close_t const *>(x.reply.decoded);
+
+						throw std::runtime_error(context + ": server channel error " + std::to_string(m->reply_code)
+								+ ", message: " + get_string(m->reply_text));
+					}
+
+					default:
+						throw std::runtime_error(context + ":  unknown server error, method id " + std::to_string( x.reply.id));
+				}
+		}
+	}
+
+	struct request
+	{
+		amqp_envelope_t envelope;
+	};
+
+	struct response
+	{
+		amqp_rpc_reply_t res;
+	};
+
+	struct service_utils
+	{
+		typedef request request_type;
+		typedef response response_type;
+
+		template <typename P, typename T>
+		auto
+		deserialize(request_type * r, P procedure, T & res) const
+		{
+
+			auto routing_key = get_string(r->envelope.routing_key);
+			auto message = get_string(r->envelope.message.body);
+			auto content_type = get_string(r->envelope.message.properties.content_type);
+			auto exchange = get_string(r->envelope.exchange);
+
+			//std::cout << "Delivery " << (unsigned) r->envelope.delivery_tag << " "
+			//		  << "exchange " << exchange << " "
+			//		  << "routingkey " << routing_key << " "
+			//		  << std::endl;
+
+			if (r->envelope.message.properties._flags & AMQP_BASIC_CONTENT_TYPE_FLAG)
+			{
+				//std::cout << "Content-type: " << content_type << std::endl;
+				//std::cout << "Message: " << message << std::endl;
+
+				iod::json_decode<typename P::route_type::parameters_type>(res, message);
+			}
+			//std::cout << "----" << std::endl;
+		}
+
+		template <typename T>
+		auto
+		serialize(response_type * r, T const & res) const
+		{
```

Comment:
I think you should serialize the responses here. It can be usefull in the case of the RPC scheme:
https://www.rabbitmq.com/tutorials/tutorial-six-python.html where the consumer (the server) has to send back a response to the client.